core/src/Cuda/Kokkos_Cuda_Parallel.hpp

```diff
@@ -1109,76 +1240,89 @@ class ParallelReduce<FunctorType, Kokkos::RangePolicy<Traits...>, ReducerType,
                                  Policy::is_graph_kernel::value ||
 #endif
                                  !std::is_same<ReducerType, InvalidType>::value;
-    if ((nwork > 0) || need_device_set) {
-      const int block_size = local_block_size(m_functor);
-
-      KOKKOS_ASSERT(block_size > 0);
-
-      m_scratch_space = cuda_internal_scratch_space(
-          m_policy.space(), ValueTraits::value_size(ReducerConditional::select(
-                                m_functor, m_reducer)) *
-                                block_size /* block_size == max block_count */);
-      m_scratch_flags =
-          cuda_internal_scratch_flags(m_policy.space(), sizeof(size_type));
-      m_unified_space = cuda_internal_scratch_unified(
-          m_policy.space(), ValueTraits::value_size(ReducerConditional::select(
-                                m_functor, m_reducer)));
-
-      // REQUIRED ( 1 , N , 1 )
-      dim3 block(1, block_size, 1);
-      // Required grid.x <= block.y
-      dim3 grid(std::min(int(block.y), int((nwork + block.y - 1) / block.y)), 1,
-                1);
 
-      // TODO @graph We need to effectively insert this in to the graph
-      const int shmem =
-          UseShflReduction
-              ? 0
-              : cuda_single_inter_block_reduce_scan_shmem<false, FunctorType,
-                                                          WorkTag>(m_functor,
-                                                                   block.y);
+#ifdef KOKKOS_ENABLE_THRUST
+    //! IsNonTrivialReduceFunctor<FunctorType>::value) {
+    if (!thrust_execute<is_thrust_possible>((nwork > 0) &&
+                                            m_result_ptr_host_accessible)) {
```

Comment:
It may be preferable to write this as `if (thrust_execute<...>(...) { return; }`, rather than nesting the rest of the body of `execute()` in this conditional.

Reply:
Agree that is better for readability, I will look into changing that